### PR TITLE
IBX-4037: Added Segment / SegmentGroup / assignment API

### PR DESCRIPTION
**Pull request by Steveb-p**
| Question      | Answer
| ------------- | ---
| JIRA Ticket   | [IBX-4037](https://issues.ibexa.co/browse/IBX-4037)
| Versions      | 4.2, 4.3.1

Adds Segment, SegmentGroup and User Segment assignment API declarations.

#### Checklist

- [ ] Text renders correctly
- [ ] Text has been checked with vale
- [ ] Description metadata is up to date
- [ ] Redirects cover removed/moved pages
- [ ] Code samples are working
- [ ] PHP code samples have been fixed with PHP CS fixer
- [ ] Added link to this PR in relevant JIRA ticket or code PR

### CODE REVIEW

File: docs/api/rest_api/rest_api_reference/input/ez-segment-groups.raml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+displayName: Segment Groups
+/{identifier}:
+    get:
+        displayName: Load segment group
+        description: Loads the given segment group
```
Comment (DominikaK) [inline docs/api/rest_api/rest_api_reference/input/ez-segment-groups.raml:5]:
```suggestion
        description: Loads the given segment group.
```

File: docs/api/rest_api/rest_api_reference/input/ez-segment-groups.raml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+                    application/vnd.ibexa.api.SegmentGroup+json:
+                        type: SegmentGroup
+                        example: !include examples/segment_groups/GET/SegmentGroup.json.example
+                    application/vnd.ibexa.api.SegmentGroup+xml:
+                        type: SegmentGroup
+                        example: !include examples/segment_groups/GET/SegmentGroup.xml.example
+            401:
+                description: Error - the user is not authorized to view this Segment Group.
+            404:
+                description: Error - no segment group with the given identifier exists.
```
Comment (DominikaK) [inline docs/api/rest_api/rest_api_reference/input/ez-segment-groups.raml:23]:
```suggestion
                description: Error - no Segment Group with the given identifier exists.
```

File: docs/api/rest_api/rest_api_reference/input/ez-types.raml [hunk 3]
```diff
@@ -1587,6 +1587,12 @@ SectionStruct:
 SectionUpdateStruct:
     description: 'This class is used to provide data for updating a Section. At least one property has to be set.'
     type: object
+Segment:
+    description: 'Segment that user can be attached to'
```
Comment (DominikaK) [inline docs/api/rest_api/rest_api_reference/input/ez-types.raml:1591]:
```suggestion
    description: 'Segment that a user can be assigned to'
```

File: docs/api/rest_api/rest_api_reference/input/ez-types.raml [hunk 4]
```diff
@@ -1587,6 +1587,12 @@ SectionStruct:
...
+    type: object
+SegmentGroup:
+    description: 'Segment group that Segment can belong to'
```
Comment (DominikaK) [inline docs/api/rest_api/rest_api_reference/input/ez-types.raml:1594]:
```suggestion
    description: 'Segment Group that a Segment can belong to'
```

File: docs/api/rest_api/rest_api_reference/input/ez-segments.raml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+                    application/vnd.ibexa.api.Segment+json:
+                        type: Segment
+                        example: !include examples/segments/GET/Segment.json.example
+                    application/vnd.ibexa.api.Segment+xml:
+                        type: Segment
+                        example: !include examples/segments/GET/Segment.xml.example
+            401:
+                description: Error - the user is not authorized to view this Segment.
+            404:
+                description: Error - no segment with the given identifier exists.
```
Comment (DominikaK) [inline docs/api/rest_api/rest_api_reference/input/ez-segments.raml:23]:
```suggestion
                description: Error - no Segment with the given identifier exists.
```

File: docs/api/rest_api/rest_api_reference/input/ez-user-users.raml [hunk 6]
```diff
@@ -303,3 +303,50 @@ head:
...
+                                type: UserSegmentList
+                                example: !include examples/user/users/user_id/segments/GET/UserSegmentsList.json.example
+                            application/vnd.ibexa.api.UserSegmentList+xml:
+                                type: UserSegmentList
+                                example: !include examples/user/users/user_id/segments/GET/UserSegmentsList.xml.example
+                    404:
+                        description: If User does not exist.
+            post:
+                displayName: Assign Segments to User
+                description: ''
```
Comment (DominikaK) [inline docs/api/rest_api/rest_api_reference/input/ez-user-users.raml:328]:
```suggestion
                description: Assign User to one or more Segments.
```
If that's correct, namely that you can assign to multiple Segments at the same time?

File: docs/api/rest_api/rest_api_reference/input/ez-user-users.raml [hunk 7]
```diff
@@ -303,3 +303,50 @@ head:
...
+                            application/vnd.ibexa.api.UserSegmentList+json:
+                                type: UserSegmentList
+                                example: !include examples/user/users/user_id/segments/GET/UserSegmentsList.json.example
+                            application/vnd.ibexa.api.UserSegmentList+xml:
+                                type: UserSegmentList
+                                example: !include examples/user/users/user_id/segments/GET/UserSegmentsList.xml.example
+                    404:
+                        description: If User does not exist.
+            post:
+                displayName: Assign Segments to User
```
Comment (DominikaK) [inline docs/api/rest_api/rest_api_reference/input/ez-user-users.raml:327]:
```suggestion
                displayName: Assign User to Segments
```
Can we go this direction?

File: docs/api/rest_api/rest_api_reference/input/ez-user-users.raml [hunk 8]
```diff
@@ -303,3 +303,50 @@ head:
...
+                        example: !include examples/user/users/user_id/segments/POST/UserSegmentAssignInput.json.example
+                responses:
+                    204:
+                        description: No Content.
+                    404:
+                        description: If Segment does not exist.
+            /{segmentIdentifier}:
+                delete:
+                    displayName: Remove Segment from User
+                    descriptions: ''
```
Comment (DominikaK) [inline docs/api/rest_api/rest_api_reference/input/ez-user-users.raml:349]:
```suggestion
                    descriptions: Unassigns User from a Segment.
```

File: docs/api/rest_api/rest_api_reference/input/ez-types.raml [hunk 9]
```diff
@@ -1996,6 +2002,15 @@ UserTokenUpdateStruct:
         time:
             description: 'Time to which the token is valid.'
             type: datetime
+
+UserSegmentList:
+    description: 'List of Segments assigned to User'
```
Comment (DominikaK) [inline docs/api/rest_api/rest_api_reference/input/ez-types.raml:2007]:
```suggestion
    description: 'List of Segments that a User is assigned to'
```

**Review by DominikaK — CHANGES_REQUESTED** [review]

File: docs/api/rest_api/rest_api_reference/input/ez-user-users.raml (hunk 6 above)
Comment (Steveb-p) [reply docs/api/rest_api/rest_api_reference/input/ez-user-users.raml:328]: That is indeed possible and correct.

**Review by DominikaK — APPROVED** [review]

**Pull request merged**